R/join_keys.R

```diff
@@ -0,0 +1,599 @@
+}
+
+#' @rdname merge_join_keys
+#' @export
+merge_join_keys <- function(join_keys_obj, new_join_keys) {
+  UseMethod("merge_join_keys", join_keys_obj)
+}
+
+#' @rdname merge_join_keys
+#' @export
+merge_join_keys.default <- function(join_keys_obj, new_join_keys) {
+  merge_join_keys(join_keys(join_keys_obj), new_join_keys)
+}
+
+#' Merging a list (or one) of `JoinKeys` objects into the current `JoinKeys` object
+#'
+#' @rdname merge_join_keys
+#'
+#' @param join_keys_obj (`JoinKeys`) object to merge the new_join_keys.
+#' @param new_join_keys  `list` of `JoinKeys` objects or single `JoinKeys` object
+#'
+#' @return a new `JoinKeys` object with the resulting merge.
+#'
+#' @export
+#'
+#' @examples
+#' jk1 <- join_keys()
+#' jk1["ds1", "ds2"] <- "some_col"
+#'
+#' jk2 <- join_keys()
+#' jk2["ds1", "ds3"] <- "new_col"
+#'
+#' merge_join_keys(jk1, jk2)
+merge_join_keys.JoinKeys <- function(join_keys_obj, new_join_keys) {
+  assert_join_keys(join_keys_obj)
+
+  if (test_join_keys(new_join_keys)) {
+    new_join_keys <- list(new_join_keys)
+  }
+
+  checkmate::assert_list(new_join_keys, types = c("JoinKeys"), min.len = 1)
+
+  result <- join_keys_obj
+
+  for (jk in new_join_keys) {
+    for (dataset_1 in names(jk)) {
+      for (dataset_2 in names(jk[[dataset_1]])) {
+        result[[dataset_1]][[dataset_2]] <- jk[[dataset_1]][[dataset_2]]
+      }
+    }
+  }
+  logger::log_trace("JoinKeys keys merged.")
+  return(result)
+}
+
+#' Prints `JoinKeys`.
+#'
+#' @inheritParams base::print
+#' @return the `x` parameter
+#'
+#' @export
+print.JoinKeys <- function(x, ...) {
+  check_ellipsis(...)
+  keys_list <- x
+  my_parents <- parents(keys_list)
+  class(keys_list) <- "list"
+  if (length(keys_list) > 0) {
+    cat(sprintf(
+      "A JoinKeys object containing foreign keys between %s datasets:\n",
+      length(keys_list)
+    ))
+    # Hide parents
+    attr(keys_list, "__parents__") <- NULL
+    print.default(keys_list)
+  } else {
+    cat("An empty JoinKeys object.")
+  }
+  invisible(x)
+}
+
+# -----------------------------------------------------------------------------
+#
+#
+# Helpers (non-exported)
+#
+
+#' Internal constructor
+#'
+#' @return an empty `JoinKeys` list
+#'
+#' @keywords internal
+new_join_keys <- function() {
+  result <- list()
+  class(result) <- c("JoinKeys", "list")
+  result
+}
+
+#' Get value of a single relationship pair
+#'
+#' @param join_keys_obj (`JoinKeys`) object that holds the relationship keys.
+#' @param dataset_1 (`character(1)`) one of the datasets to retrieve keys (
+#' order of the datasets is irrelevant).
+#' @param dataset_2 (`character(1)`) the other dataset to retrieve keys (the
+#' order of the datasets is irrelevant).
+#'
+#' @return Character vector with keys or (if one of the datasets is omitted) a
+#' list of relationship pairs. If both datasets are omitted it returens the
+#' `JoinKeys` object
+#'
+#' @keywords internal
+get_join_key <- function(join_keys_obj, dataset_1 = NULL, dataset_2 = NULL) {
```

Comment:
Is it needed? I think `[[` is enough

Reply:
Removed.

I wanted to keep the getter and setters as much as functions as possible.

However, that is not possible when adding `[[<-.JoinKey` custom code.